gb restore fails with "too many open files"

```
2022/01/03 09:02:35 open /root/skycacherestore/skycache-1640113240/51716,27413.zst for write
panic: open /root/skycacherestore/skycache-1640113240/51716,27413.zst: too many open files

goroutine 1 [running]:
github.com/leijurv/gb/download.execute(0xc000bd5660, 0x20, 0x20, 0x0, 0xc02794b9b0, 0xc02d57b6b0, 0xc02794b9e0)
	github.com/leijurv/gb/download/restore.go:273 +0xc93
github.com/leijurv/gb/download.Restore(0xc010a3dff0, 0xf, 0xc0001263d8, 0x15, 0x0)
	github.com/leijurv/gb/download/restore.go:249 +0x16f9
main.main.func16(0xc0000ea160, 0xc00010e300, 0xc00010e330)
	github.com/leijurv/gb/main.go:302 +0xab
github.com/urfave/cli.HandleAction(0xec66e0, 0x109f0e0, 0xc0000ea160, 0xc0000ea160, 0x0)
	github.com/urfave/cli@v1.22.5/app.go:524 +0x105
github.com/urfave/cli.Command.Run(0x10302f9, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x103fa2c, 0x15, 0x0, ...)
	github.com/urfave/cli@v1.22.5/command.go:173 +0x579
github.com/urfave/cli.(*App).Run(0xc0001ad500, 0xc0000320c0, 0x4, 0x4, 0x0, 0x0)
	github.com/urfave/cli@v1.22.5/app.go:277 +0x808
main.main()
	github.com/leijurv/gb/main.go:373 +0x1385
```
the entire restore was expected to be about 1.8 million files

--- a/download/restore.go
+++ b/download/restore.go
@@ -263,7 +263,8 @@ func execute(rest Restoration) {
 		paths = append(paths, path)
 	}
 	// To avoid potentially exhausting the open file limit, write to up to 500 files at a time.
-	// This will do multiple downloads but this is only likely to happen with the 0 byte file or a small file
+	// After the first 500 files are restored, one will be chosen as the source of data for the next 500 files.
+	diskSource := rest.nominatedSource
 	for i := 0; i < len(paths); i += 500 {
 		chunk := paths[i:min(len(rest.destinations), i+500)]
 		handles := make([]*os.File, 0)
@@ -298,12 +299,12 @@ func execute(rest Restoration) {
 		out = io.MultiWriter(out, &hs)
 
 		var src io.Reader
-		if rest.nominatedSource == nil {
+		if diskSource == nil {
 			log.Println("Fetching from storage")
 			src = CatEz(rest.hash)
 		} else {
-			log.Println("Reading locally, from", *rest.nominatedSource)
-			f, err := os.Open(*rest.nominatedSource)
+			log.Println("Reading locally, from", *diskSource)
+			f, err := os.Open(*diskSource)
 			if err != nil {
 				panic(err)
 			}
@@ -318,6 +319,7 @@ func execute(rest Restoration) {
 			panic("wrong")
 		}
 		log.Println("Success")
+		diskSource = &chunk[0]
 
 		for _, f := range handles {
 			f.Close()

--- a/download/restore.go
+++ b/download/restore.go
@@ -250,57 +250,79 @@ func Restore(src string, dest string, timestamp int64) {
 	}
 }
 
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+	return y
+}
+
 func execute(rest Restoration) {
-	writers := make([]io.Writer, 0)
-	for path, item := range rest.destinations {
-		dir := filepath.Dir(path)
-		mode := item.permissions
+	paths := make([]string, 0)
+	for path, _ := range rest.destinations {
+		paths = append(paths, path)
+	}
+	// To avoid potentially exhausting the open file limit, write to up to 500 files at a time.
+	// This will do multiple downloads but this is only likely to happen with the 0 byte file or a small file
+	for i := 0; i < len(paths); i += 500 {
+		chunk := paths[i:min(len(rest.destinations), i+500)]
+		handles := make([]*os.File, 0)
+		writers := make([]io.Writer, 0)
+		for _, path := range chunk {
+			dir := filepath.Dir(path)
+			mode := rest.destinations[path].permissions
 
-		// https://stackoverflow.com/a/31151508/2277831
-		dirMode := mode               // start with perms of the file
-		dirMode |= (mode >> 2) & 0111 // for group and other, allow execute (dir read) if they can read
-		dirMode |= 0700               // we must have full access no matter what, otherwise this recursive mkdir won't work in the first place
+			// https://stackoverflow.com/a/31151508/2277831
+			dirMode := mode               // start with perms of the file
+			dirMode |= (mode >> 2) & 0111 // for group and other, allow execute (dir read) if they can read
+			dirMode |= 0700               // we must have full access no matter what, otherwise this recursive mkdir won't work in the first place
 
-		log.Println("mkdir", dir, "with original", mode, "overridden to", dirMode)
-		err := os.MkdirAll(dir, dirMode)
-		if err != nil {
-			panic(err)
+			log.Println("mkdir", dir, "with original", mode, "overridden to", dirMode)
+			err := os.MkdirAll(dir, dirMode)
+			if err != nil {
+				panic(err)
+			}
+
+			log.Println("open", path, "for write")
+			f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, mode)
+			if err != nil {
+				panic(err)
+			}
+			handles = append(handles, f)
+			writers = append(writers, f)
 		}
 
-		log.Println("open", path, "for write")
-		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, mode)
-		if err != nil {
-			panic(err)
-		}
-		defer f.Close()
-		writers = append(writers, f)
-	}
-	out := io.MultiWriter(writers...)
+		out := io.MultiWriter(writers...)
 
-	hs := utils.NewSHA256HasherSizer()
-	out = io.MultiWriter(out, &hs)
+		hs := utils.NewSHA256HasherSizer()
+		out = io.MultiWriter(out, &hs)
 
-	var src io.Reader
-	if rest.nominatedSource == nil {
-		log.Println("Fetching from storage")
-		src = CatEz(rest.hash)
-	} else {
-		log.Println("Reading locally, from", *rest.nominatedSource)
-		f, err := os.Open(*rest.nominatedSource)
-		if err != nil {
-			panic(err)
+		var src io.Reader
+		if rest.nominatedSource == nil {
+			log.Println("Fetching from storage")
+			src = CatEz(rest.hash)
+		} else {
+			log.Println("Reading locally, from", *rest.nominatedSource)
+			f, err := os.Open(*rest.nominatedSource)
+			if err != nil {
+				panic(err)
+			}
+			defer f.Close()
+			src = f
 		}
-		defer f.Close()
-		src = f
+		utils.Copy(out, src)
+		log.Println("Expecting size and hash:", rest.size, hex.EncodeToString(rest.hash))
+		hash, size := hs.HashAndSize()
+		log.Println("Got size and hash:", size, hex.EncodeToString(hash))
+		if size != rest.size || !bytes.Equal(hash, rest.hash) {
+			panic("wrong")
+		}
+		log.Println("Success")
+
+		for _, f := range handles {
+			f.Close()
+		}
 	}
-	utils.Copy(out, src)
-	log.Println("Expecting size and hash:", rest.size, hex.EncodeToString(rest.hash))
-	hash, size := hs.HashAndSize()
-	log.Println("Got size and hash:", size, hex.EncodeToString(hash))
-	if size != rest.size || !bytes.Equal(hash, rest.hash) {
-		panic("wrong")
-	}
-	log.Println("Success")
 }
 
 func statSources(plan map[[32]byte]*Restoration) {

